analytics/analytics/management/commands/upload_build_timings.py

```diff
@@ -0,0 +1,111 @@
+import json
+import os
+import re
+import tempfile
+import zipfile
+
+import gitlab
+from gitlab.v4.objects import Project, ProjectJob
+
+from analytics.models import Job, Phase, Timer
+
+# Grab env vars
+GITLAB_TOKEN = os.environ["GITLAB_TOKEN"]
+JOB_INPUT_DATA = os.environ["JOB_INPUT_DATA"]
+
+# Other constants
+PACKAGE_NAME_REGEX = r"(.+)/[a-zA-Z0-9]+ .+"
+MICROARCHS = {"x86_64_v4", "x86_64_v3", "x86_64_v2", "graviton3", "graviton2"}
+
+
+# Instantiate gitlab api wrapper
+gl = gitlab.Gitlab("https://gitlab.spack.io", GITLAB_TOKEN)
+
+
+def create_job(project: Project, job: ProjectJob) -> Job:
+    # Grab package name and runner tags
+    package_name = re.match(PACKAGE_NAME_REGEX, job.name).group(1)
+    runner_tags = gl.runners.get(job.runner["id"]).tag_list
+
+    # Get microarch
+    microarch = ""
+    for tag in job.tag_list:
+        if tag in MICROARCHS:
+            microarch = tag
+            break
+
+    # Return created job
+    return Job.objects.create(
+        job_id=job.get_id(),
+        project_id=project.get_id(),
+        name=job.name,
+        started_at=job.started_at,
+        duration=job.duration,
+        ref=job.ref,
+        package_name=package_name,
+        aws=("aws" in runner_tags),
+        microarch=microarch,
+    )
+
+
+def get_timings_json(job: ProjectJob):
+    # Download job artifacts and parse timings json
+    with tempfile.NamedTemporaryFile(suffix=".zip") as temp:
+        artifacts_file = temp.name
+        with open(artifacts_file, "wb") as f:
+            job.artifacts(streamed=True, action=f.write)
+
+        # Read in timing json
+        timing_filename = "jobs_scratch_dir/user_data/install_times.json"
+        with zipfile.ZipFile(artifacts_file) as zfile:
+            with zfile.open(timing_filename) as timing_file:
+                timings_json = json.load(timing_file)
+
+    return timings_json
+
+
+def main():
+    # Read input data and extract params
+    job_input_data = json.loads(JOB_INPUT_DATA)
+    job_id = job_input_data["build_id"]
+
+    # Retrieve project and job from gitlab API
+    gl_project = gl.projects.get(job_input_data["project_id"])
+    gl_job = gl_project.jobs.get(job_input_data["build_id"])
+
+    # Get or create job record
+    job = Job.objects.filter(job_id=job_id).first()
+    if job is None:
+        job = create_job(gl_project, gl_job)
+
+    # Iterate through each timer and create timers and phase results
+    phases = []
+    for entry in get_timings_json(gl_job):
+        timer = Timer.objects.create(
+            name=entry["name"],
+            hash=entry["hash"],
```

Comment:
some timers are not going to have `hash` or `cache` since they are not related to building packages. We should get `entry.get("hash", None)` to set these to null

Reply:
This has been fixed.